warehouse/malware/checks/setup_patterns/check.py

```
@@ -0,0 +1,120 @@
+                for name in zipobj.namelist():
+                    parts = os.path.split(name)
+                    if len(parts) == 2 and parts[1] == "setup.py":
+                        return zipobj.read(name)
+                return None
+        else:
+            # NOTE: is_zipfile doesn't rewind the fileobj it's given.
+            archive_stream.seek(0)
+
+            # NOTE: We don't need to perform a sanity check on
+            # the (presumed) tarfile's compression here, since we're
+            # extracting from a stream that's already gone through
+            # upload validation.
+            # See _is_valid_dist_file in forklift/legacy.py.
+            try:
+                with tarfile.open(fileobj=archive_stream) as tarobj:
+                    member = tarobj.next()
+                    while member:
+                        parts = os.path.split(member.name)
+                        if len(parts) == 2 and parts[1] == "setup.py":
+                            return tarobj.extractfile(member).read()
+                        member = tarobj.next()
+                    return None
+            except tarfile.TarError:
+                return None
+
+    def scan(self, **kwargs):
+        file_id = kwargs.get("obj_id")
+        file_url = kwargs.get("file_url")
+        if file_id is None or file_url is None:
+            # TODO: Maybe raise here, since the absence of these
+            # arguments is a use/user error.
+            return
+
+        file = self.db.query(File).get(file_id)
+        if file.packagetype != "sdist":
+            # Per PEP 491: bdists do not contain setup.py.
+            # This check only scans dists that contain setup.py, so
+            # we have nothing to perform.
+            return
+
+        # TODO(ww): Handle connection errors, timeouts?
+        # The check harness will catch everything eventually,
+        # so these might not be worth handling explicitly.
```

I agree with the approach outlined in this comment. Maybe remove the `file_request.ok` checking and let it raise an `AttributeError` a few lines down on `file_request.content` to that `run_check` retries a few times.

---

Sounds good! I'm refactoring this into a general utility but I'll make it as noisy as possible (in terms of exceptions) to ensure that we propagate them outwards for retrying.